exchanges/request/request.go

```diff
@@ -1,475 +1,208 @@
+	for k, v := range i.Headers {
 		req.Header.Add(k, v)
 	}
 
-	if r.UserAgent != "" && req.Header.Get("User-Agent") == "" {
-		req.Header.Add("User-Agent", r.UserAgent)
+	if r.UserAgent != "" && req.Header.Get(userAgent) == "" {
+		req.Header.Add(userAgent, r.UserAgent)
 	}
 
 	return req, nil
 }
 
 // DoRequest performs a HTTP/HTTPS request with the supplied params
-func (r *Requester) DoRequest(req *http.Request, path string, body io.Reader, result interface{}, authRequest, verbose, httpDebug, httpRecord bool) error {
-	if verbose {
+func (r *Requester) DoRequest(req *http.Request, p *Item) error {
+	if p.Verbose {
 		log.Debugf(log.Global,
-			"%s exchange request path: %s requires rate limiter: %v",
+			"%s exchange request path: %s",
 			r.Name,
-			path,
-			r.RequiresRateLimiter())
+			p.Path)
 
 		for k, d := range req.Header {
-			log.Debugf(log.Global, "%s exchange request header [%s]: %s", r.Name, k, d)
+			log.Debugf(log.Global,
+				"%s exchange request header [%s]: %s",
+				r.Name,
+				k,
+				d)
 		}
 		log.Debugf(log.Global,
-			"%s exchange request type: %s", r.Name, req.Method)
-		log.Debugf(log.Global,
-			"%s exchange request body: %v", r.Name, body)
+			"%s exchange request type: %s",
+			r.Name,
+			req.Method)
+
+		if p.Body != nil {
+			log.Debugf(log.Global,
+				"%s exchange request body: %v",
+				r.Name,
+				p.Body)
+		}
 	}
 
 	var timeoutError error
 	for i := 0; i < r.timeoutRetryAttempts+1; i++ {
+		// Initiate a rate limit reservation and sleep on requested endpoint
+		err := r.InitiateRateLimit(p.Endpoint)
+		if err != nil {
+			return err
+		}
+
 		resp, err := r.HTTPClient.Do(req)
 		if err != nil {
 			if timeoutErr, ok := err.(net.Error); ok && timeoutErr.Timeout() {
-				if verbose {
-					log.Errorf(log.ExchangeSys, "%s request has timed-out retrying request, count %d",
+				if p.Verbose {
+					log.Errorf(log.ExchangeSys,
+						"%s request has timed-out retrying request, count %d",
 						r.Name,
 						i)
 				}
 				timeoutError = err
 				continue
 			}
-
-			if r.RequiresRateLimiter() {
-				r.DecrementRequests(authRequest)
-			}
 			return err
 		}
-		if resp == nil {
-			if r.RequiresRateLimiter() {
-				r.DecrementRequests(authRequest)
-			}
-			return errors.New("resp is nil")
-		}
 
-		var reader io.ReadCloser
-		switch resp.Header.Get("Content-Encoding") {
-		case "gzip":
-			reader, err = gzip.NewReader(resp.Body)
-			defer reader.Close()
-			if err != nil {
-				return err
-			}
-
-		case "json":
-			reader = resp.Body
-
-		default:
-			switch {
-			case strings.Contains(resp.Header.Get("Content-Type"), "application/json"):
-				reader = resp.Body
-
-			default:
-				if verbose {
-					log.Warnf(log.ExchangeSys,
-						"%s request response content type differs from JSON; received %v [path: %s]\n",
-						r.Name,
-						resp.Header.Get("Content-Type"),
-						path)
-				}
-				reader = resp.Body
+		if !strings.Contains(resp.Header.Get("Content-Type"), "application/json") {
```

Comment:
Is this needed anymore? doesn't really serve any purpose as we just continue and either the unmarshal later on is going to fail if they are not sending valid json or nothing is going to happen